pkg/certificateprovider/azurekeyvault/provider.go

```diff
@@ -0,0 +1,368 @@
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get vault")
+	}
+	logrus.Infof("vaultName %v, vaultURL %v", keyvaultName, *vaultURL)
+
+	kvClient, err := initializeKvClient(ctx, azureCloudEnv.KeyVaultEndpoint, azureCloudEnv.ActiveDirectoryEndpoint, tenantID, workloadIdentityClientID)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get keyvault client")
+	}
+
+	// 3. for each object , get content bytes
+	files := []types.CertificateFile{}
+	for _, keyVaultCert := range keyVaultCerts {
+		logrus.Infof("fetching object from key vault, certName %v,  keyvault %v", keyVaultCert.CertificateName, keyvaultName)
+
+		resolvedKvCerts, err := resolveCertificateVersions(ctx, kvClient, keyVaultCert, *vaultURL)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, resolvedKvCert := range resolvedKvCerts {
+			// fetch the object from Key Vault
+			result, err := getCertificate(ctx, kvClient, *vaultURL, resolvedKvCert)
+			if err != nil {
+				return nil, err
+			}
+
+			for idx := range result {
+				r := result[idx]
+				objectContent := []byte(r.content)
+
+				file := types.CertificateFile{
+					Path:    resolvedKvCert.GetFileName(),
+					Content: objectContent,
+					Version: r.version,
+				}
+
+				files = append(files, file)
+				logrus.Infof("added file %v to response file", file.Path)
+			}
+		}
+	}
+	return files, nil
+}
+
+// formatKeyVaultCertificate formats the fields in KeyVaultCertificate
+func formatKeyVaultCertificate(object *types.KeyVaultCertificate) {
+	if object == nil {
+		return
+	}
+	objectPtr := reflect.ValueOf(object)
+	objectValue := objectPtr.Elem()
+
+	for i := 0; i < objectValue.NumField(); i++ {
+		field := objectValue.Field(i)
+		if field.Type() != reflect.TypeOf("") {
+			continue
+		}
+		str := field.Interface().(string)
+		str = strings.TrimSpace(str)
+		field.SetString(str)
+	}
+}
+
+// ParseAzureEnvironment returns azure environment by name
+func ParseAzureEnvironment(cloudName string) (*azure.Environment, error) {
+	var env azure.Environment
+	var err error
+	if cloudName == "" {
+		env = azure.PublicCloud
+	} else {
+		env, err = azure.EnvironmentFromName(cloudName)
+	}
+	return &env, err
+}
+
+func getVaultURL(keyvaultName string, KeyVaultDNSSuffix string) (vaultURL *string, err error) {
+	// Key Vault name must be a 3-24 character string
+	if len(keyvaultName) < 3 || len(keyvaultName) > 24 {
+		return nil, errors.Errorf("Invalid vault name: %q, must be between 3 and 24 chars", keyvaultName)
+	}
+	// See docs for validation spec: https://docs.microsoft.com/en-us/azure/key-vault/about-keys-secrets-and-certificates#objects-identifiers-and-versioning
+	isValid := regexp.MustCompile(`^[-A-Za-z0-9]+$`).MatchString
+	if !isValid(keyvaultName) {
+		return nil, errors.Errorf("Invalid vault name: %q, must match [-a-zA-Z0-9]{3,24}", keyvaultName)
+	}
+
+	vaultDNSSuffixValue := KeyVaultDNSSuffix
+	vaultURI := "https://" + keyvaultName + "." + vaultDNSSuffixValue + "/"
+	return &vaultURI, nil
+}
```

Comment:
Instead of constructing the vault url would it be easier to get just get keyvault URL as input from the CRD instead of keyvault name?

Reply:
I think from the customer perspective, name of keyvault is more  user friendly, what do you think?

Reply:
renamed to vaultUri to match the property name from azure keyvault, thanks!